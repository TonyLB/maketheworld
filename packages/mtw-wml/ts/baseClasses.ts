import { ParseTag } from "./parser/baseClasses";

type SchemaNestingBase<T extends SchemaTag> = {
    contents: T[];
}

export type LegalAssetContents = SchemaActionTag | SchemaComputedTag | SchemaConditionTag | SchemaExitTag | SchemaFeatureTag | SchemaImageTag | SchemaImportTag | SchemaMapTag | SchemaRoomTag | SchemaVariableTag
export type LegalConditionContents = SchemaActionTag | SchemaComputedTag | SchemaConditionTag | SchemaExitTag | SchemaFeatureTag | SchemaImageTag | SchemaImportTag | SchemaMapTag | SchemaRoomTag | SchemaVariableTag

type SchemaAssetBase = {
    key: string;
    fileName?: string;
    zone?: string;
    subFolder?: string;
    player?: string;
} & SchemaNestingBase<LegalAssetContents>

export type SchemaAssetTag = {
    tag: 'Asset';
} & SchemaAssetBase

export type SchemaStoryTag = {
    tag: 'Story';
    instance: boolean;
} & SchemaAssetBase

export type SchemaImageTag = {
    tag: 'Image';
    key: string;
    fileURL?: string;
}

type SchemaPronouns = {
    subject: string;
    object: string;
    possessive: string;
    adjective: string;
    reflexive: string;
}

export type SchemaCharacterTag = {
    tag: 'Character';
    player?: string;
    name: string;
    pronouns: SchemaPronouns;
    firstImpression?: string;
    oneCoolThing?: string;
    outfit?: string;
    image?: SchemaImageTag;
}

export type SchemaVariableTag = {
    tag: 'Variable';
    key: string;
    default?: string;
}

export type SchemaComputedTag = {
    tag: 'Computed';
    key: string;
    src: string;
    dependencies: string[];
}

export type SchemaActionTag = {
    tag: 'Action';
    src: string;
}

export type SchemaUseTag = {
    tag: 'Use';
    key: string;
    as?: string;
    type?: string;
}

type SchemaImportMapping = {
    key: string;
    type: 'Room' | 'Feature' | 'Variable' | 'Computed' | 'Action' | 'Map'
}

export type SchemaImportTag = {
    tag: 'Import';
    from: string;
    mapping: Record<string, SchemaImportMapping>;
}

//
// TODO:  Figure out how to define limits on schemaConditionTag contents
// without causing circular reference
//
export type SchemaConditionTag = {
    tag: 'Condition';
    if: string;
    dependencies: string[];
} & SchemaNestingBase<any>

export type SchemaExitTag = {
    tag: 'Exit';
    key: string;
    name: string;
    to: string;
    from: string;
}

export type SchemaLinkTag = {
    tag: 'Link';
    to: string;
} & SchemaNestingBase<SchemaStringTag>

export type SchemaDescriptionTag = {
    tag: 'Description';
    display: 'before' | 'after' | 'replace';
    spaceBefore: boolean;
    spaceAfter: boolean;
} & SchemaNestingBase<SchemaStringTag | SchemaLinkTag | SchemaLineBreakTag>

export type SchemaLineBreakTag = {
    tag: 'br';
}

export type SchemaNameTag = {
    tag: 'Name';
    name: string;
}

export type SchemaRoomTag = {
    tag: 'Room';
    key: string;
    global: boolean;
    display?: string;
    x?: number;
    y?: number;
} & SchemaNestingBase<SchemaDescriptionTag | SchemaExitTag | SchemaFeatureTag | SchemaNameTag>

export type SchemaFeatureTag = {
    tag: 'Feature';
    key: string;
    global: boolean;
}& SchemaNestingBase<SchemaDescriptionTag | SchemaNameTag>

export type SchemaMapTag = {
    tag: 'Map';
    key: string;
} & SchemaNestingBase<SchemaExitTag | SchemaImageTag | SchemaRoomTag>

export type SchemaStringTag = {
    tag: 'String';
    value: string;
}

export type SchemaTag = SchemaAssetTag |
    SchemaStoryTag |
    SchemaCharacterTag |
    SchemaImageTag |
    SchemaVariableTag |
    SchemaComputedTag |
    SchemaActionTag |
    SchemaUseTag |
    SchemaImportTag |
    SchemaConditionTag |
    SchemaExitTag |
    SchemaDescriptionTag |
    SchemaLineBreakTag |
    SchemaLinkTag |
    SchemaNameTag |
    SchemaRoomTag |
    SchemaFeatureTag |
    SchemaMapTag |
    SchemaStringTag

export class SchemaException extends Error {
    parseTag: ParseTag;
    constructor(message: string, parseTag: ParseTag) {
        super(message)
        this.name = 'WMLSchemaException'
        this.parseTag = parseTag
    }
}

export const isSchemaString = (value: SchemaTag): value is SchemaStringTag => (value.tag === 'String')
