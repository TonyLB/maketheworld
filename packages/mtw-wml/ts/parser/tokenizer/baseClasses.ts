import SourceStream from "./sourceStream";

export type TokenBase = {
    source: string;
    startIdx: number;
    endIdx: number;
}

export type TokenComment = {
    type: 'Comment'
} & TokenBase

export type TokenWhitespace = {
    type: 'Whitespace'
} & TokenBase

export type TokenExpressionValue = {
    type: 'ExpressionValue'
} & TokenBase

export type TokenKeyValue = {
    type: 'KeyValue';
    key: string;
} & TokenBase

export type TokenLiteralValue = {
    type: 'LiteralValue';
    value: string;
} & TokenBase

export type TokenProperty = {
    type: 'Property'
} & TokenBase

export type TokenBeginTagOpen = {
    type: 'BeginTagOpen'
} & TokenBase

export type TokenBeginTagClose = {
    type: 'BeginTagClose';
} & TokenBase

export type TokenEndTag = {
    type: 'EndTag';
} & TokenBase

export type TokenDescription = {
    type: 'Description';
} & TokenBase

export type TokenError = {
    type: 'Error';
    message: string;
} & TokenBase

export type Token = TokenComment |
    TokenWhitespace |
    TokenExpressionValue |
    TokenKeyValue |
    TokenLiteralValue |
    TokenProperty |
    TokenBeginTagOpen |
    TokenBeginTagClose |
    TokenEndTag |
    TokenDescription |
    TokenError

export type Tokenizer<T extends TokenBase> = {
    (sourceStream: SourceStream): T | TokenError | undefined
}

export const isTokenError = (item: TokenBase & { type: string }): item is TokenError => (item.type === 'Error')
