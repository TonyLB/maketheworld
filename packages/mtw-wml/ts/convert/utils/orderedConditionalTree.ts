import { deepEqual } from "../../lib/objects";
import { NormalConditionStatement } from "../../normalize/baseClasses"
import { SchemaTag } from "../../schema/baseClasses"

export type OrderedConditionalNode = {
    conditions: NormalConditionStatement[];
    contents: OrderedConditionalTree;
}

export type OrderedConditionalTree = (SchemaTag | OrderedConditionalNode)[]

export type FlattenedConditionalNode = {
    conditions: NormalConditionStatement[];
    contents: SchemaTag[];
}

const isConditionNode = (value: SchemaTag | OrderedConditionalNode): value is OrderedConditionalNode => (!('tag' in value))

const lastNode = (list: FlattenedConditionalNode[]): FlattenedConditionalNode | undefined => {
    if (list.length === 0) {
        return undefined
    }
    else {
        return list.slice(-1)[0]
    }
}

//
// flattenOrderedConditionalTree transforms an OrderedConditionalTree to a functionally equivalent FlattenedConditionalNode
// by depth-first walking the tree and outputting each different set of conditions as its own FlattenedConditionalNode in order
//
export const flattenOrderedConditionalTreeReducer = (previous: FlattenedConditionalNode[], context: NormalConditionStatement[], tree: OrderedConditionalTree): FlattenedConditionalNode[] => {
    return tree.reduce<FlattenedConditionalNode[]>((accumulator, node) => {
        if (isConditionNode(node)) {
            return flattenOrderedConditionalTreeReducer(
                accumulator,
                [ ...context, ...node.conditions ],
                node.contents
            )
        }
        else {
            const previousItem = lastNode(accumulator)
            if (previousItem && deepEqual(previousItem.conditions, context)) {
                return [
                    ...(previous.slice(0, -1)),
                    {
                        conditions: context,
                        contents: [
                            ...previousItem.contents,
                            node
                        ]
                    }
                ]
            }
            else {
                return [
                    ...accumulator,
                    {
                        conditions: context,
                        contents: [node]
                    }
                ]
            }
        }
    }, previous)
}

export const flattenOrderedConditionalTree = (tree: OrderedConditionalTree): FlattenedConditionalNode[] => (flattenOrderedConditionalTreeReducer([], [], tree))
