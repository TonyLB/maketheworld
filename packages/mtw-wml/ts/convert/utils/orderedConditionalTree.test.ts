import { flattenOrderedConditionalTree, mergeOrderedConditionalTrees, navigationSequence, unflattenOrderedConditionalTree } from './orderedConditionalTree'
import { makeSchemaTag } from './index'
import { SchemaConditionTag, SchemaTag } from '../../schema/baseClasses'

describe('orderedConditionalTree', () => {
    describe('flattenOrderedConditionalTree', () => {
        it('should return an empty list from an empty tree', () => {
            expect(flattenOrderedConditionalTree('Description')([])).toEqual([])
        })

        it('should flatten a nested tree', () => {
            const testTree: SchemaTag[] = [
                makeSchemaTag({
                    tag: 'If',
                    contextTag: 'Description',
                    conditions: [{
                        if: 'test',
                        dependencies: ['test'],
                    }],
                    contents: [
                        makeSchemaTag({
                            tag: 'If',
                            contextTag: 'Description',        
                            conditions: [{
                                if: 'test3',
                                not: true,
                                dependencies: ['test3']
                            },
                            {
                                if: 'test2',
                                dependencies: ['test2']
                            }],
                            contents: [makeSchemaTag({
                                tag: 'String',
                                key: 'NestedWithThreeConditions'
                            })]
                        }),
                        makeSchemaTag({ tag: 'String' as 'String', key: 'NestedWithOneCondition' })
                    ]
                }),
                makeSchemaTag({ tag: 'String', value: 'NotNested'}),
                makeSchemaTag({
                    tag: 'If',
                    contextTag: 'Description',
                    conditions: [{
                        if: 'test4',
                        dependencies: ['test4']
                    }],
                    contents: [makeSchemaTag({ tag: 'String', key: 'SecondNestedWithOneCondition'})]
                })
            ]
            expect(flattenOrderedConditionalTree('Description')(testTree)).toMatchSnapshot()
        })

        it('should merge adjacent items with similar conditions', () => {
            const testTree: SchemaTag[] = [
                makeSchemaTag({
                    tag: 'If',
                    contextTag: 'Description',
                    conditions: [{
                        if: 'test',
                        dependencies: ['test'],
                    }],
                    contents: [
                        makeSchemaTag({
                            tag: 'If',
                            contextTag: 'Description',        
                            conditions: [{
                                if: 'test2',
                                dependencies: ['test2']
                            }],
                            contents: [makeSchemaTag({
                                tag: 'String',
                                key: 'TestOne'
                            })]
                        })
                    ]
                }),
                makeSchemaTag({
                    tag: 'If',
                    contextTag: 'Description',
                    conditions: [{
                        if: 'test',
                        dependencies: ['test']
                    },
                    {
                        if: 'test2',
                        dependencies: ['test2']
                    }],
                    contents: [
                        makeSchemaTag({ tag: 'String', key: 'TestTwo'}),
                        makeSchemaTag({ tag: 'String', key: 'TestThree'})
                    ]
                })
            ]
            expect(flattenOrderedConditionalTree('Description')(testTree)).toMatchSnapshot()
        })

    })

    describe('unflattenOrderedConditionalTree', () => {
        it('should return an empty tree from an empty list', () => {
            expect(unflattenOrderedConditionalTree([])).toEqual([])
        })

        it('should unflatten a list', () => {
            const condition1 = {
                if: 'test',
                dependencies: ['test']
            }
            const condition2 = {
                if: 'test2',
                dependencies: ['test2']
            }
            const condition3 = {
                if: 'test3',
                dependencies: ['test3']
            }
            const testList: SchemaConditionTag[] = [
                makeSchemaTag({ tag: 'If', contextTag: 'Description', conditions: [condition1, condition2], contents: [makeSchemaTag({ tag: 'String', value: 'TestA' }), makeSchemaTag({ tag: 'String', value: 'TestB' })] }),
                makeSchemaTag({ tag: 'If', contextTag: 'Description', conditions: [condition1, condition2], contents: [makeSchemaTag({ tag: 'String', value: 'TestC' })] }),
                makeSchemaTag({ tag: 'If', contextTag: 'Description', conditions: [condition1, condition2, condition3], contents: [makeSchemaTag({ tag: 'String', value: 'TestD' })] }),
                makeSchemaTag({ tag: 'If', contextTag: 'Description', conditions: [condition1, condition3], contents: [makeSchemaTag({ tag: 'String', value: 'TestE'})] }),
                makeSchemaTag({ tag: 'If', contextTag: 'Description', conditions: [], contents: [makeSchemaTag({ tag: 'String', value: 'TestF' })] })
            ] as SchemaConditionTag[]
            expect(unflattenOrderedConditionalTree(testList)).toMatchSnapshot()
        })

        it('should collapse deep and unbranching conditional paths', () => {
            const condition1 = {
                if: 'test',
                dependencies: ['test']
            }
            const condition2 = {
                if: 'test2',
                dependencies: ['test2']
            }
            const condition3 = {
                if: 'test3',
                dependencies: ['test3']
            }
            const testList = [
                makeSchemaTag({ tag: 'If', contextTag: 'Description', conditions: [condition1, condition2, condition3], contents: [makeSchemaTag({ tag: 'String', value: 'TestA' }), makeSchemaTag({ tag: 'String', value: 'TestB' })] }),
                makeSchemaTag({ tag: 'If', contextTag: 'Description', conditions: [condition1, condition2], contents: [makeSchemaTag({ tag: 'String', value: 'TestC' })] }),
                makeSchemaTag({ tag: 'If', contextTag: 'Description', conditions: [condition1, condition2, condition3], contents: [makeSchemaTag({ tag: 'String', value: 'TestD' })] }),
                makeSchemaTag({ tag: 'If', contextTag: 'Description', conditions: [], contents: [makeSchemaTag({ tag: 'String', value: 'TestE' })] })
            ] as SchemaConditionTag[]
            expect(unflattenOrderedConditionalTree(testList)).toMatchSnapshot()
        })

    })

    describe('navigationSequence', () => {
        it('should return a traverse down and up on a single location', () => {
            expect(navigationSequence([[1, 2, 3]])).toEqual([[], [1], [1, 2], [1, 2, 3], [1, 2], [1], []])
        })
    })

    describe('mergeOrderedConditionalTrees', () => {
        it('should return unchanged on merger with empty tree', () => {
            const testTree: SchemaTag[] = [
                makeSchemaTag({ tag: 'String', key: 'TestZero'}),
                makeSchemaTag({
                    tag: 'If',
                    contextTag: 'Description',
                    conditions: [{
                        if: 'test',
                        dependencies: ['test'],
                    }],
                    contents: [makeSchemaTag({
                        tag: 'String',
                        key: 'TestOne'
                    })]
                }),
                makeSchemaTag({
                    tag: 'If',
                    contextTag: 'Description',
                    conditions: [{
                        if: 'test2',
                        dependencies: ['test2']
                    },
                    {
                        if: 'test',
                        dependencies: ['test']
                    }],
                    contents: [
                        makeSchemaTag({ tag: 'String', key: 'TestTwo'}),
                        makeSchemaTag({ tag: 'String', key: 'TestThree'})
                    ]
                })
            ]
            expect(mergeOrderedConditionalTrees('Description')(testTree, [])).toEqual(testTree)
            expect(mergeOrderedConditionalTrees('Description')([], testTree)).toEqual(testTree)
        })

        it('should merge differently ordered trees', () => {
            const testTreeOne: SchemaTag[] = [
                makeSchemaTag({ tag: 'String', key: 'TestZero'}),
                makeSchemaTag({
                    tag: 'If',
                    contextTag: 'Description',
                    conditions: [{
                        if: 'test',
                        dependencies: ['test'],
                    }],
                    contents: [makeSchemaTag({
                        tag: 'String',
                        value: 'TestOne'
                    })]
                }),
                makeSchemaTag({
                    tag: 'If',
                    contextTag: 'Description',
                    conditions: [{
                        if: 'test2',
                        dependencies: ['test2']
                    },
                    {
                        if: 'test',
                        dependencies: ['test']
                    }],
                    contents: [
                        makeSchemaTag({ tag: 'String', value: 'TestTwo'}),
                        makeSchemaTag({ tag: 'String', value: 'TestThree'})
                    ]
                })
            ]
            const testTreeTwo: SchemaTag[] = [
                makeSchemaTag({ tag: 'String', value: 'TestFour'}),
                makeSchemaTag({
                    tag: 'If',
                    contextTag: 'Description',
                    conditions: [{
                        if: 'test2',
                        dependencies: ['test2']
                    }],
                    contents: [makeSchemaTag({
                        tag: 'If',
                        contextTag: 'Description',
    
                        conditions: [{
                            if: 'test3',
                            dependencies: ['test3']
                        }],
                        contents: [makeSchemaTag({ tag: 'String', value: 'TestFive'})]
                    }),
                    makeSchemaTag({
                        tag: 'If',
                        contextTag: 'Description',
                            conditions: [{
                            if: 'test',
                            dependencies: ['test']
                        }],
                        contents: [makeSchemaTag({ tag: 'String', value: 'TestSix'})]

                    })]
                }),
                makeSchemaTag({
                    tag: 'If',
                    contextTag: 'Description',
                    conditions: [{
                        if: 'test',
                        dependencies: ['test'],
                    }],
                    contents: [makeSchemaTag({
                        tag: 'String',
                        value: 'TestSeven'
                    })]
                })
            ]
            expect(mergeOrderedConditionalTrees('Description')(testTreeOne, testTreeTwo)).toMatchSnapshot()
            expect(mergeOrderedConditionalTrees('Description')(testTreeTwo, testTreeOne)).toMatchSnapshot()
        })

    })
})
