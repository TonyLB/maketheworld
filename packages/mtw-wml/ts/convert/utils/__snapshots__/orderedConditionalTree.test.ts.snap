// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`orderedConditionalTree flattenOrderedConditionalTree should flatten a nested tree 1`] = `
Array [
  Object {
    "conditions": Array [
      Object {
        "dependencies": Array [
          "test",
        ],
        "if": "test",
      },
      Object {
        "dependencies": Array [
          "test3",
        ],
        "if": "test3",
        "not": true,
      },
      Object {
        "dependencies": Array [
          "test2",
        ],
        "if": "test2",
      },
    ],
    "contents": Array [
      Object {
        "key": "NestedWithThreeConditions",
        "parse": Object {
          "endTagToken": 0,
          "startTagToken": 0,
          "tag": "Space",
        },
        "tag": "String",
      },
    ],
    "contextTag": "Description",
    "parse": Object {
      "endTagToken": 0,
      "startTagToken": 0,
      "tag": "Space",
    },
    "tag": "If",
  },
  Object {
    "conditions": Array [
      Object {
        "dependencies": Array [
          "test",
        ],
        "if": "test",
      },
    ],
    "contents": Array [
      Object {
        "key": "NestedWithOneCondition",
        "parse": Object {
          "endTagToken": 0,
          "startTagToken": 0,
          "tag": "Space",
        },
        "tag": "String",
      },
    ],
    "contextTag": "Description",
    "parse": Object {
      "endTagToken": 0,
      "startTagToken": 0,
      "tag": "Space",
    },
    "tag": "If",
  },
  Object {
    "conditions": Array [],
    "contents": Array [
      Object {
        "parse": Object {
          "endTagToken": 0,
          "startTagToken": 0,
          "tag": "Space",
        },
        "tag": "String",
        "value": "NotNested",
      },
    ],
    "contextTag": "Description",
    "parse": Object {
      "endTagToken": 0,
      "startTagToken": 0,
      "tag": "Space",
    },
    "tag": "If",
  },
  Object {
    "conditions": Array [
      Object {
        "dependencies": Array [
          "test4",
        ],
        "if": "test4",
      },
    ],
    "contents": Array [
      Object {
        "key": "SecondNestedWithOneCondition",
        "parse": Object {
          "endTagToken": 0,
          "startTagToken": 0,
          "tag": "Space",
        },
        "tag": "String",
      },
    ],
    "contextTag": "Description",
    "parse": Object {
      "endTagToken": 0,
      "startTagToken": 0,
      "tag": "Space",
    },
    "tag": "If",
  },
]
`;

exports[`orderedConditionalTree flattenOrderedConditionalTree should merge adjacent items with similar conditions 1`] = `
Array [
  Object {
    "conditions": Array [
      Object {
        "dependencies": Array [
          "test",
        ],
        "if": "test",
      },
      Object {
        "dependencies": Array [
          "test2",
        ],
        "if": "test2",
      },
    ],
    "contents": Array [
      Object {
        "key": "TestOne",
        "parse": Object {
          "endTagToken": 0,
          "startTagToken": 0,
          "tag": "Space",
        },
        "tag": "String",
      },
      Object {
        "key": "TestTwo",
        "parse": Object {
          "endTagToken": 0,
          "startTagToken": 0,
          "tag": "Space",
        },
        "tag": "String",
      },
      Object {
        "key": "TestThree",
        "parse": Object {
          "endTagToken": 0,
          "startTagToken": 0,
          "tag": "Space",
        },
        "tag": "String",
      },
    ],
    "contextTag": "Description",
    "parse": Object {
      "endTagToken": 0,
      "startTagToken": 0,
      "tag": "Space",
    },
    "tag": "If",
  },
]
`;

exports[`orderedConditionalTree mergeOrderedConditionalTrees should merge differently ordered trees 1`] = `
Array [
  Object {
    "key": "TestZero",
    "parse": Object {
      "endTagToken": 0,
      "startTagToken": 0,
      "tag": "Space",
    },
    "tag": "String",
  },
  Object {
    "parse": Object {
      "endTagToken": 0,
      "startTagToken": 0,
      "tag": "Space",
    },
    "tag": "String",
    "value": "TestFour",
  },
  Object {
    "conditions": Array [
      Object {
        "dependencies": Array [
          "test",
        ],
        "if": "test",
      },
    ],
    "contents": Array [
      Object {
        "parse": Object {
          "endTagToken": 0,
          "startTagToken": 0,
          "tag": "Space",
        },
        "tag": "String",
        "value": "TestOne",
      },
    ],
    "contextTag": "Description",
    "parse": Object {
      "endTagToken": 0,
      "startTagToken": 0,
      "tag": "Space",
    },
    "tag": "If",
  },
  Object {
    "conditions": Array [
      Object {
        "dependencies": Array [
          "test2",
        ],
        "if": "test2",
      },
    ],
    "contents": Array [
      Object {
        "conditions": Array [
          Object {
            "dependencies": Array [
              "test3",
            ],
            "if": "test3",
          },
        ],
        "contents": Array [
          Object {
            "parse": Object {
              "endTagToken": 0,
              "startTagToken": 0,
              "tag": "Space",
            },
            "tag": "String",
            "value": "TestFive",
          },
        ],
        "contextTag": "Description",
        "parse": Object {
          "endTagToken": 0,
          "startTagToken": 0,
          "tag": "Space",
        },
        "tag": "If",
      },
      Object {
        "conditions": Array [
          Object {
            "dependencies": Array [
              "test",
            ],
            "if": "test",
          },
        ],
        "contents": Array [
          Object {
            "parse": Object {
              "endTagToken": 0,
              "startTagToken": 0,
              "tag": "Space",
            },
            "tag": "String",
            "value": "TestTwo",
          },
          Object {
            "parse": Object {
              "endTagToken": 0,
              "startTagToken": 0,
              "tag": "Space",
            },
            "tag": "String",
            "value": "TestThree",
          },
          Object {
            "parse": Object {
              "endTagToken": 0,
              "startTagToken": 0,
              "tag": "Space",
            },
            "tag": "String",
            "value": "TestSix",
          },
        ],
        "contextTag": "Description",
        "parse": Object {
          "endTagToken": 0,
          "startTagToken": 0,
          "tag": "Space",
        },
        "tag": "If",
      },
    ],
    "contextTag": "Description",
    "parse": Object {
      "endTagToken": 0,
      "startTagToken": 0,
      "tag": "Space",
    },
    "tag": "If",
  },
  Object {
    "conditions": Array [
      Object {
        "dependencies": Array [
          "test",
        ],
        "if": "test",
      },
    ],
    "contents": Array [
      Object {
        "parse": Object {
          "endTagToken": 0,
          "startTagToken": 0,
          "tag": "Space",
        },
        "tag": "String",
        "value": "TestSeven",
      },
    ],
    "contextTag": "Description",
    "parse": Object {
      "endTagToken": 0,
      "startTagToken": 0,
      "tag": "Space",
    },
    "tag": "If",
  },
]
`;

exports[`orderedConditionalTree mergeOrderedConditionalTrees should merge differently ordered trees 2`] = `
Array [
  Object {
    "parse": Object {
      "endTagToken": 0,
      "startTagToken": 0,
      "tag": "Space",
    },
    "tag": "String",
    "value": "TestFour",
  },
  Object {
    "key": "TestZero",
    "parse": Object {
      "endTagToken": 0,
      "startTagToken": 0,
      "tag": "Space",
    },
    "tag": "String",
  },
  Object {
    "conditions": Array [
      Object {
        "dependencies": Array [
          "test",
        ],
        "if": "test",
      },
    ],
    "contents": Array [
      Object {
        "parse": Object {
          "endTagToken": 0,
          "startTagToken": 0,
          "tag": "Space",
        },
        "tag": "String",
        "value": "TestOne",
      },
    ],
    "contextTag": "Description",
    "parse": Object {
      "endTagToken": 0,
      "startTagToken": 0,
      "tag": "Space",
    },
    "tag": "If",
  },
  Object {
    "conditions": Array [
      Object {
        "dependencies": Array [
          "test2",
        ],
        "if": "test2",
      },
    ],
    "contents": Array [
      Object {
        "conditions": Array [
          Object {
            "dependencies": Array [
              "test3",
            ],
            "if": "test3",
          },
        ],
        "contents": Array [
          Object {
            "parse": Object {
              "endTagToken": 0,
              "startTagToken": 0,
              "tag": "Space",
            },
            "tag": "String",
            "value": "TestFive",
          },
        ],
        "contextTag": "Description",
        "parse": Object {
          "endTagToken": 0,
          "startTagToken": 0,
          "tag": "Space",
        },
        "tag": "If",
      },
      Object {
        "conditions": Array [
          Object {
            "dependencies": Array [
              "test",
            ],
            "if": "test",
          },
        ],
        "contents": Array [
          Object {
            "parse": Object {
              "endTagToken": 0,
              "startTagToken": 0,
              "tag": "Space",
            },
            "tag": "String",
            "value": "TestSix",
          },
          Object {
            "parse": Object {
              "endTagToken": 0,
              "startTagToken": 0,
              "tag": "Space",
            },
            "tag": "String",
            "value": "TestTwo",
          },
          Object {
            "parse": Object {
              "endTagToken": 0,
              "startTagToken": 0,
              "tag": "Space",
            },
            "tag": "String",
            "value": "TestThree",
          },
        ],
        "contextTag": "Description",
        "parse": Object {
          "endTagToken": 0,
          "startTagToken": 0,
          "tag": "Space",
        },
        "tag": "If",
      },
    ],
    "contextTag": "Description",
    "parse": Object {
      "endTagToken": 0,
      "startTagToken": 0,
      "tag": "Space",
    },
    "tag": "If",
  },
  Object {
    "conditions": Array [
      Object {
        "dependencies": Array [
          "test",
        ],
        "if": "test",
      },
    ],
    "contents": Array [
      Object {
        "parse": Object {
          "endTagToken": 0,
          "startTagToken": 0,
          "tag": "Space",
        },
        "tag": "String",
        "value": "TestSeven",
      },
    ],
    "contextTag": "Description",
    "parse": Object {
      "endTagToken": 0,
      "startTagToken": 0,
      "tag": "Space",
    },
    "tag": "If",
  },
]
`;

exports[`orderedConditionalTree unflattenOrderedConditionalTree should collapse deep and unbranching conditional paths 1`] = `
Array [
  Object {
    "conditions": Array [
      Object {
        "dependencies": Array [
          "test",
        ],
        "if": "test",
      },
      Object {
        "dependencies": Array [
          "test2",
        ],
        "if": "test2",
      },
    ],
    "contents": Array [
      Object {
        "conditions": Array [
          Object {
            "dependencies": Array [
              "test3",
            ],
            "if": "test3",
          },
        ],
        "contents": Array [
          Object {
            "parse": Object {
              "endTagToken": 0,
              "startTagToken": 0,
              "tag": "Space",
            },
            "tag": "String",
            "value": "TestA",
          },
          Object {
            "parse": Object {
              "endTagToken": 0,
              "startTagToken": 0,
              "tag": "Space",
            },
            "tag": "String",
            "value": "TestB",
          },
        ],
        "contextTag": "Description",
        "parse": Object {
          "endTagToken": 0,
          "startTagToken": 0,
          "tag": "Space",
        },
        "tag": "If",
      },
      Object {
        "parse": Object {
          "endTagToken": 0,
          "startTagToken": 0,
          "tag": "Space",
        },
        "tag": "String",
        "value": "TestC",
      },
      Object {
        "conditions": Array [
          Object {
            "dependencies": Array [
              "test3",
            ],
            "if": "test3",
          },
        ],
        "contents": Array [
          Object {
            "parse": Object {
              "endTagToken": 0,
              "startTagToken": 0,
              "tag": "Space",
            },
            "tag": "String",
            "value": "TestD",
          },
        ],
        "contextTag": "Description",
        "parse": Object {
          "endTagToken": 0,
          "startTagToken": 0,
          "tag": "Space",
        },
        "tag": "If",
      },
    ],
    "contextTag": "Description",
    "parse": Object {
      "endTagToken": 0,
      "startTagToken": 0,
      "tag": "Space",
    },
    "tag": "If",
  },
  Object {
    "parse": Object {
      "endTagToken": 0,
      "startTagToken": 0,
      "tag": "Space",
    },
    "tag": "String",
    "value": "TestE",
  },
]
`;

exports[`orderedConditionalTree unflattenOrderedConditionalTree should unflatten a list 1`] = `
Array [
  Object {
    "conditions": Array [
      Object {
        "dependencies": Array [
          "test",
        ],
        "if": "test",
      },
    ],
    "contents": Array [
      Object {
        "conditions": Array [
          Object {
            "dependencies": Array [
              "test2",
            ],
            "if": "test2",
          },
        ],
        "contents": Array [
          Object {
            "parse": Object {
              "endTagToken": 0,
              "startTagToken": 0,
              "tag": "Space",
            },
            "tag": "String",
            "value": "TestA",
          },
          Object {
            "parse": Object {
              "endTagToken": 0,
              "startTagToken": 0,
              "tag": "Space",
            },
            "tag": "String",
            "value": "TestB",
          },
          Object {
            "parse": Object {
              "endTagToken": 0,
              "startTagToken": 0,
              "tag": "Space",
            },
            "tag": "String",
            "value": "TestC",
          },
          Object {
            "conditions": Array [
              Object {
                "dependencies": Array [
                  "test3",
                ],
                "if": "test3",
              },
            ],
            "contents": Array [
              Object {
                "parse": Object {
                  "endTagToken": 0,
                  "startTagToken": 0,
                  "tag": "Space",
                },
                "tag": "String",
                "value": "TestD",
              },
            ],
            "contextTag": "Description",
            "parse": Object {
              "endTagToken": 0,
              "startTagToken": 0,
              "tag": "Space",
            },
            "tag": "If",
          },
        ],
        "contextTag": "Description",
        "parse": Object {
          "endTagToken": 0,
          "startTagToken": 0,
          "tag": "Space",
        },
        "tag": "If",
      },
      Object {
        "conditions": Array [
          Object {
            "dependencies": Array [
              "test3",
            ],
            "if": "test3",
          },
        ],
        "contents": Array [
          Object {
            "parse": Object {
              "endTagToken": 0,
              "startTagToken": 0,
              "tag": "Space",
            },
            "tag": "String",
            "value": "TestE",
          },
        ],
        "contextTag": "Description",
        "parse": Object {
          "endTagToken": 0,
          "startTagToken": 0,
          "tag": "Space",
        },
        "tag": "If",
      },
    ],
    "contextTag": "Description",
    "parse": Object {
      "endTagToken": 0,
      "startTagToken": 0,
      "tag": "Space",
    },
    "tag": "If",
  },
  Object {
    "parse": Object {
      "endTagToken": 0,
      "startTagToken": 0,
      "tag": "Space",
    },
    "tag": "String",
    "value": "TestF",
  },
]
`;
